relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -1,152 +1,228 @@
-                    }
-                }
-            }
-        }
+/// Trait implemented by providers of fields (Events and Trace Contexts).
+/// The fields will be used by rules to check if they apply.
+trait FieldValueProvider {
+    /// gets the value of a field
+    fn get_value(&self, path: &str) -> FieldValue;
+    /// what type of rule can be applied to this provider
+    fn get_rule_type(&self) -> RuleType;
+}
 
-        // match against the project
-        if !self.project_ids.is_empty() && !self.project_ids.contains(&project_id) {
-            return false;
+impl FieldValueProvider for Event {
+    fn get_value(&self, field_name: &str) -> FieldValue {
+        match field_name {
+            "event.release" => match self.release.as_str() {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s),
+            },
+            "event.environment" => match self.environment.as_str() {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s),
+            },
+            "event.user_segment" => FieldValue::None, // Not available at this time
+            _ => FieldValue::None,
         }
-
-        // match against the release
-        if !self.releases.is_empty() {
-            match release {
-                None => return false,
-                Some(ref release) => {
-                    if !self.releases.is_match(release) {
-                        return false;
-                    }
-                }
+    }
+    fn get_rule_type(&self) -> RuleType {
+        if let Some(ty) = self.ty.value() {
+            if *ty == EventType::Transaction {
+                return RuleType::Transaction;
             }
         }
+        RuleType::Error
+    }
+}
 
-        // match against the user_segment
-        if !self.user_segments.is_empty() {
-            match user_segment {
-                None => return false,
-                Some(ref user_segment) => {
-                    if !self.user_segments.contains(user_segment) {
-                        return false;
-                    }
-                }
-            }
+impl FieldValueProvider for TraceContext {
+    fn get_value(&self, field_name: &str) -> FieldValue {
+        match field_name {
+            "trace.release" => match &self.release {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s.as_ref()),
+            },
+            "trace.environment" => match &self.environment {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s.as_ref()),
+            },
+            "trace.user_segment" => match &self.user_segment {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s.as_ref()),
+            },
+            _ => FieldValue::None,
         }
-        true
+    }
+    fn get_rule_type(&self) -> RuleType {
+        RuleType::Trace
+    }
+}
+
+//TODO make this more efficient
```

Comment:
Are you still planning to resolve this TODO?

Reply:
left from before we used the unicase crate, removing